### DISCUSSION TITLE
/articles/entra-id-validating-generated-oauth-tokens/

### DISCUSSION THREAD

**Discussion by giscus** [Voitanos/voitanos.io-comments#225] [Comments]
In this post, Andrew Connell explains how you can validate OAuth tokens generated by Entra ID in your custom apps.
https://www.andrewconnell.com/articles/entra-id-validating-generated-oauth-tokens/

**Comment by lanorix** [comment]
Great write-up.
I have a question though. How does this work for azure B2C tenants? we have dev/qual/prod that we are trying to accomplish this with and the document does not seem to target a specific tenant to get the keys?  We would like to verify the signature of the access token that is given by the app registration within a B2C tenant.

**Reply by andrewconnell** [reply]
Good question! You should be able to get the keys using the TID of the tenant you're targeting. Just swap out the `common` part of the URL for your tenant's ID. 

Although I just did that for my tenant and the keys look identical for what was used by the common endpoint.

**Comment by norgie** [comment]
Andrew, you wrote "What you’re looking for is a key that has the same thumbprint of the x.509 certificate (SHA-1 thumbprint) that’s listed in the header of the JWT token you got from Entra ID". So how do you get hold of this JWT if it is issued by AAD/Entra ID itself in connection with e.g. provisioning a SCIM endpoint (e.g. implented in C#). The doc says that if you leave "the Secret Token field blank, Microsoft Entra ID includes an OAuth bearer token issued from Microsoft Entra ID with each request. Apps that use Microsoft Entra ID as an identity provider can validate this Microsoft Entra ID-issued token." So how do you set up your rest-api or any other web-api to validate this before you even get the token? Seems like a chicken-and-egg situation to me.

**Reply by andrewconnell** [reply]
@norgie said:

> The doc says that if you leave "the Secret Token field blank, Microsoft Entra ID includes an OAuth bearer token issued from Microsoft Entra ID with each request. Apps that use Microsoft Entra ID as an identity provider can validate this Microsoft Entra ID-issued token." ...

What doc? Sorry... the text you're quoting isn't from my article so I'm not sure what you're referring to.

Also, can't comment on provisioning a SCIM endpoint... I haven't done that before.

My article is about validating a JWT token you obtain for Microsoft Entra ID, so my assumption is you already have the token.

**Reply by norgie** [reply]
My apologies. Didn't pull my head out of the context it was in before posting.

**Comment by jawadcsc84** [comment]
Hello Andrew, I tried to come up with c# equivalent for the above code. but its throwing error while converting base64 format. below the c# code. Am i missing something here...

```csharp
private const string KeyString = "-----BEGIN CERTIFICATE-----\nMIIC==...\n-----END CERTIFICATE-----";
public static object DecodeValidToken(string accessToken) {
  var keyBytes = Convert.FromBase64String(KeyString);
  var certificate = new X509Certificate2(keyBytes);
  var validationParameters = new TokenValidationParameters {
    ValidateIssuerSigningKey = true,
    IssuerSigningKey = new X509SecurityKey(certificate),
    ValidateIssuer = false,
    ValidateAudience = false
  };

  var tokenHandler = new JwtSecurityTokenHandler();
  SecurityToken validatedToken;
  var principal = tokenHandler.ValidateToken(accessToken, validationParameters, out validatedToken);
  return principal.Claims; // or return principal for more details
}
```

**Reply by andrewconnell** [reply]
Sorry... not familiar with doing this with C#/.NET. But... 

The key isn't Base64 encoded... the *access* token is Base64 encoded. In your code, you're decoding the key which isn't necessary.

**Comment by rrakso** [comment]
I'm using this solution based on [jwks-rsa](https://www.npmjs.com/package/jwks-rsa) and [passport-jwt](https://www.npmjs.com/package/passport-jwt) packages to fetch always fresh certificate: 
```typescript
secretOrKeyProvider: passportJwtSecret({
    cache: true,
    rateLimit: true,
    jwksRequestsPerMinute: 5,
    jwksUri: `https://${authority}/${tenantId}/discovery/${version}/keys`,
}),
```
Where: _tenantId:_ `9debb24a-d069-4715-9845-054796f0267e`, _clientId:_ `14229163-1776-4f0f-8eb3-f4a9120c9258`, _authority:_ `login.microsoftonline.com`, _version:_ `v2.0`

**Reply by DakotaWray2** [reply]
Is there a Microsoft built library to validate tokens server-side these days? There are some older recommendations for passport-azure-ad but it's deprecated with the readme saying they are working on something. But that seems to be from years ago.

**Reply by andrewconnell** [reply]
@DakotaWray2 I _literally_ was just made aware of this the [jwt-validate](https://www.npmjs.com/package/jwt-validate) yesterday that helps with some of this stuff. I plan to review it and possibly incorporate it into this article soon. It's _very_ new so YMMV, but it seems to be intended to address some of these things.

My feedback to the author was the package description needs a lot more info explaining what it does. In talking to him, it uses [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken) and [jwks-rsa](https://www.npmjs.com/package/jwks-rsa) and adds a bunch of additional helpers, plus it has some Microsoft Entra ID-specific helpers.

I haven't used it yet, but might be worth evaluating.

**Reply by DakotaWray2** [reply]
@andrewconnell Thanks! I will check it out.

**Comment by andrewconnell** [comment]
I just posted a big refresh to this article this week... if you've read this page before, there's a lot more to it now!

**Comment by mhstudioos** [comment]
Hi, great article with lots of insights.

Anyone already able to successfully perform the above in php?

**Reply by andrewconnell** [reply]
You can definitely do this with PHP... or any language. I'm not experienced with PHP, but I bet you could use any of the AI tools out there to convert what I've shared in this post to another language.

**Reply by mhstudioos** [reply]
Hi thanks for the reply, 

i've tried but i think it got the wrong token to validate. the header contains more properties as your example:

```
{
  "typ": "JWT",
  "nonce": "0H0gQZiQgs2Q2y6---",
  "alg": "RS256",
  "x5t": "CNv0OI3Rwq----",
  "kid": "CNv0OI3Rwq-----"
}
```

**Reply by andrewconnell** [reply]
It's hard to help without seeing the token, or at least enough of it to have some context.

The snippet I showed in my article, specifically the picture with the label **Locate the key ID (`kid`) property**, is only the header portion of the JWT access token. An access token has 3 segments, as I explained in the section **[Validating JWT tokens were created by Microsoft Entra ID](https://www.voitanos.io/blog/validating-entra-id-generated-oauth-tokens/#validating-jwt-tokens-were-created-by-microsoft-entra-id)**:

- header
- payload
- signature

The screenshot I referenced earlier in this comment only showed the header portion... that's all you need to figure out which public key you need to use to validate the token.

**Comment by n-3rd23** [comment]
I tried to validate the access token but I'm getting invalid signature. In my case I am using workforce tenant. the iss claim I'm getting is sts.windows.net

**Reply by n-3rd23** [reply]
found the answer here : https://xsreality.medium.com/making-azure-ad-oidc-compliant-5734b70c43ff
